packages/compiler/core/checker.ts

```diff
@@ -1789,6 +1960,1115 @@ export function createChecker(program: Program): Checker {
+    }
+
+    return lastVal;
+  }
+
+  function evalProjectionArithmeticExpression(
+    node: ProjectionArithmeticExpressionNode
+  ): StringLiteralType | ReturnRecord {
+    const lhs = evalProjectionNode(node.left);
+    if (lhs.kind === "Return") {
+      return lhs;
+    }
+    const rhs = evalProjectionNode(node.right);
+    if (rhs.kind === "Return") {
+      return rhs;
+    }
+
+    // todo: handle numbers
+    return createType({
+      kind: "String",
+      value: (lhs as StringLiteralType).value + (rhs as StringLiteralType).value,
+      node: node as any,
+    });
+  }
+
+  function evalProjectionStatement(
+    node: ProjectionNode,
+    target: Type,
+    args: (Type | number | string | boolean)[]
+  ): Type {
+    const parentProjection = node.parent! as ProjectionStatementNode;
+    if (target.projections.indexOf(parentProjection) === -1) {
+      throw new Error(`Can't apply projection ${parentProjection.id.sv} to type ${target.kind}`);
+    }
+    let topLevelProjection = false;
+    if (!currentProjectionDirection) {
+      topLevelProjection = true;
+      currentProjectionDirection = node.direction;
+    }
+    const originalContext = evalContext;
+    evalContext = createEvalContext(node);
+    for (const [i, param] of node.parameters.entries()) {
+      if (!args[i]) {
+        throw new Error("need argument for parameter " + node.parameters[i]);
+      }
+
+      let argVal = args[i];
+      let typeVal;
+
+      if (typeof argVal === "number" || typeof argVal === "string" || typeof argVal === "boolean") {
+        typeVal = valueToLiteralType(argVal);
+      } else {
+        typeVal = argVal;
+      }
+
+      evalContext.locals.set(param.id.sv, typeVal);
+    }
+    const clone = cloneType(target);
+
+    clone.projectionSource = target;
+    evalContext.locals.set("self", clone);
+    let lastVal: TypeOrReturnRecord = voidType;
+    for (const item of node.body) {
+      lastVal = evalProjectionNode(item);
+      if (lastVal.kind === "Return") {
+        break;
+      }
+    }
+
+    if (topLevelProjection) {
+      currentProjectionDirection = undefined;
+    }
+    evalContext = originalContext;
+
+    if (lastVal.kind === "Return") {
+      return lastVal.value;
+    } else {
+      return clone;
+    }
+  }
+
+  function evalProjectionExpressionStatement(node: ProjectionExpressionStatement) {
+    return evalProjectionNode(node.expr);
+  }
+
+  function evalProjectionCallExpression(node: ProjectionCallExpressionNode) {
+    const target = evalProjectionNode(node.target);
+
+    if (!target) throw new Error("target undefined");
+    const args = [];
+    for (const arg of node.arguments) {
+      args.push(evalProjectionNode(arg));
+    }
+
+    if (target.kind !== "Function") {
+      throw new Error("Can't call non-function, got type " + target.kind);
+    }
+
+    return target.call(...args);
+  }
+
+  function evalProjectionMemberExpression(node: ProjectionMemberExpressionNode): Type {
```

Comment:
This function is ridiculous but I'm not entirely sure the best way to refactor it yet. But, given a member expression, this function finds the base type and manufactures the corresponding members to expose to projections. IOW, this function implements the built-in APIs (since we don't want to just blindly expose everything in JS to cadl and vise versa).

Reply:
At the very least put these in a data structure that's enumerable.